src/main/java/com/exactpro/th2/act/BiDirectionalMessageReceiver.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright 2021-2021 Exactpro (Exactpro Systems Limited)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exactpro.th2.act;
+
+import static com.google.protobuf.TextFormat.shortDebugString;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.exactpro.th2.common.grpc.Direction;
+import com.exactpro.th2.common.grpc.Message;
+import com.exactpro.th2.common.grpc.MessageBatch;
+import com.exactpro.th2.common.grpc.MessageID;
+import com.exactpro.th2.common.schema.message.MessageListener;
+
+public class BiDirectionalMessageReceiver extends AbstractMessageReceiver {
+
+    private enum State {
+        START, OUTGOING_MATCHED, INCOMING_MATCHED;
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(BiDirectionalMessageReceiver.class);
+
+    private final SubscriptionManager subscriptionManager;
+    private final CheckRule outgoingRule;
+    private final Function<Message, CheckRule> incomingRuleSupplier;
+
+    private final Queue<MessageBatch> incomingBuffer = new LinkedList<>();
```

Comment:
We integrate via full collection, maybe use Collection as the variable type and ArrayList as implementation

Reply:
We use the buffer as a FIFO queue. So, I have decided to use LinkedList here. We can use ArrayDeque instead for example

Reply:
Ok, let's continue to use LinkedList.